### PR TITLE
add sets support

### PR DESCRIPTION
**Pull request by Lakate**
Add [sets](https://github.com/bem/bem-sdk/pull/213) support 

For each set run multiple webpack builds in parallel (using parallel-webpack)
Each set builds in ```appBuild/setName/```
If no set static files will be in ```appBuild/main/```

Run webpackDevServer with command  ```npm run start setName``` 
if no set ```npm run start```

### CODE REVIEW

File: packages/bem-react-scripts/config/webpack.config.dev.js [hunk 1]
```diff
@@ -74,7 +77,7 @@ module.exports = {
     // This does not produce a real file. It's just the virtual path that is
     // served by WebpackDevServer in development. This is the JS bundle
     // containing code from all our entry points, and the Webpack runtime.
-    filename: 'static/js/bundle.js',
+    filename: `static/js/${setName}.bundle.js`,
```
Comment (qfox) [inline packages/bem-react-scripts/config/webpack.config.dev.js]: Looks like we can avoid `main.bundle.js`

File: packages/bem-react-scripts/scripts/build.js [hunk 2]
```diff
@@ -68,132 +68,18 @@ function build(previousFileSizes) {
 
   let compiler;
   try {
-    compiler = webpack(config);
+    compiler = parallelWebpack.run(
```
Comment (qfox) [inline packages/bem-react-scripts/scripts/build.js]: Looks like we don't need compiler variable

File: packages/bem-react-scripts/config/webpack.config.dev.js [hunk 3]
```diff
@@ -232,6 +235,7 @@ module.exports = {
     new HtmlWebpackPlugin({
       inject: true,
       template: paths.appHtml,
+      filename: `${paths.appBuild}/${setName}/index.html`,
```
Comment (qfox) [inline packages/bem-react-scripts/config/webpack.config.dev.js]: and here?

File: packages/bem-react-scripts/scripts/build.js [hunk 4]
```diff
@@ -68,132 +68,18 @@ function build(previousFileSizes) {
...
+        configPath,
+        {
+            watch: false,
+            maxRetries: 1,
+            stats: true // defaults to false
```
Comment (qfox) [inline packages/bem-react-scripts/scripts/build.js]: Better to use ENV for this value

Comment (qfox) [reply packages/bem-react-scripts/scripts/build.js]: Prob better to use params like in parallel-webpack: https://github.com/trivago/parallel-webpack#specifying-retry-limits

File: README.md [hunk 5]
```diff
@@ -117,10 +117,19 @@ Your app is ready to be deployed!
...
 * [bem-react-core](https://github.com/bem/bem-react-core)
 * [raw-loader](https://github.com/webpack-contrib/raw-loader) includes content of *.inline.* files
 
+## Usage sets
+
+if you want to make several bundles and have their own implementation of blocks, you need to use sets
+
+1. You need specify the required sets in bem.config
+2. Then build the project. You can specify some options for [parallel-webpack](https://github.com/trivago/parallel-webpack) (watch, maxRetries, stats) Example: ```npm run build --watch=true```
+3. Finally, run app with a ```npm run start setName```
```
Comment (qfox) [inline README.md]: `npm run start [platform]`?

File: packages/bem-react-scripts/config/webpack.config.dev.js [hunk 6]
```diff
@@ -74,7 +77,7 @@ module.exports = {
     // This does not produce a real file. It's just the virtual path that is
     // served by WebpackDevServer in development. This is the JS bundle
     // containing code from all our entry points, and the Webpack runtime.
-    filename: 'static/js/bundle.js',
+    filename: setName ? `static/js/${setName}.bundle.js` : `static/js/[name].bundle.js`,
```
Comment (qfox) [inline packages/bem-react-scripts/config/webpack.config.dev.js:80]: this is still differ from what it was?

Comment (Lakate) [reply packages/bem-react-scripts/config/webpack.config.dev.js:80]: now if not setName then the build will work as usual

File: README.md [hunk 7]
```diff
@@ -117,10 +117,19 @@ Your app is ready to be deployed!
 Currently it includes something more than original Create React App package:
 
 * [webpack-bem-loader](https://github.com/bem/webpack-bem-loader)
+* [bem/sdk.config](https://github.com/bem/bem-sdk/tree/master/packages/config)
```
Comment (Vittly) [inline README.md]: ... to handle `.bemrc`

File: README.md [hunk 8]
```diff
@@ -117,10 +117,19 @@ Your app is ready to be deployed!
...
 * [babel-plugin-bem-import](https://github.com/bem/babel-plugin-bem-import)
 * [bem-react-core](https://github.com/bem/bem-react-core)
 * [raw-loader](https://github.com/webpack-contrib/raw-loader) includes content of *.inline.* files
 
+## Usage sets
```
Comment (Vittly) [inline README.md]: Let's write about feature not tool -> `## Building multiple bundles`

File: README.md [hunk 9]
```diff
@@ -117,10 +117,19 @@ Your app is ready to be deployed!
...
+
+if you want to make several bundles and have their own implementation of blocks, you need to use sets
```
Comment (Vittly) [inline README.md]: Please start with a capital letter

Comment (Vittly) [reply README.md]: Add link to sets definition at bem-sdk/config readme

File: README.md [hunk 10]
```diff
@@ -117,10 +117,19 @@ Your app is ready to be deployed!
...
+
+1. You need specify the required sets in bem.config
```
Comment (Vittly) [inline README.md]: in `.bemrc`

File: README.md [hunk 11]
```diff
@@ -117,10 +117,19 @@ Your app is ready to be deployed!
...
+2. Then build the project. You can specify some options for [parallel-webpack](https://github.com/trivago/parallel-webpack) (watch, maxRetries, stats) Example: ```npm run build --watch=true```
```
Comment (Vittly) [inline README.md]: You can delete "Example: ..." it is not informative

File: README.md [hunk 12]
```diff
@@ -117,10 +117,19 @@ Your app is ready to be deployed!
...
+3. Finally, run app with a ```npm run start platform```
```
Comment (Vittly) [inline README.md]: not platform -> `npm run start <set-name>`

File: README.md [hunk 13]
```diff
@@ -117,10 +117,19 @@ Your app is ready to be deployed!
...
+
```
Comment (Vittly) [inline README.md]: I like shorter texts like:

* Add sets in `.bemrc`
* Build. Here you can specify some options ...
* Run with `npm start ...`

File: packages/bem-react-scripts/config/webpack.config.dev.js [hunk 14]
```diff
@@ -33,6 +33,9 @@ const publicUrl = '';
 // Get environment variables to inject into our app.
 const env = getClientEnvironment(publicUrl);
 
+// Get platform name, after start app like `npm run start platform`
+const platform = process.argv[2];
```
Comment (Vittly) [inline packages/bem-react-scripts/config/webpack.config.dev.js]: may be `setName` is better

File: packages/bem-react-scripts/scripts/build.js [hunk 15]
```diff
@@ -66,134 +76,19 @@ function printErrors(summary, errors) {
...
   } catch (err) {
     printErrors('Failed to compile.', [err]);
     process.exit(1);
   }
-
-  compiler.run((err, stats) => {
-    if (err) {
-      printErrors('Failed to compile.', [err]);
-      process.exit(1);
-    }
```
Comment (Vittly) [inline packages/bem-react-scripts/scripts/build.js:81]: We delete a lot of colourful messages :(

File: packages/bem-react-scripts/scripts/build.js [hunk 16]
```diff
@@ -23,8 +23,18 @@ const chalk = require('chalk');
 const fs = require('fs-extra');
 const path = require('path');
 const url = require('url');
-const webpack = require('webpack');
-const config = require('../config/webpack.config.prod');
+
+const parallelWebpack = require('parallel-webpack');
+const webpackArgv = require('minimist')(process.argv.slice(2), {
```
Comment (Vittly) [inline packages/bem-react-scripts/scripts/build.js]: parallelWebpackArgv

File: packages/bem-react-scripts/config/webpack.config.prod.js [hunk 17]
```diff
@@ -42,261 +42,273 @@ if (env.stringified['process.env'].NODE_ENV !== '"production"') {
...
+                },
+                // Process JS with Babel.
+                {
+                    test: /\.(js|jsx)$/,
+                    use: [
+                        {
+                            loader: 'webpack-bem-loader',
+                            options: {
+                                levels: paths.appLevels,
+                                sets: setName,
```
Comment (Vittly) [inline packages/bem-react-scripts/config/webpack.config.prod.js]: Lets call option `set`

File: packages/bem-react-scripts/config/webpack.config.prod.js [hunk 18]
```diff
@@ -42,261 +42,273 @@ if (env.stringified['process.env'].NODE_ENV !== '"production"') {
...
-      // ** ADDING/UPDATING LOADERS **
-      // The "url" loader handles all assets unless explicitly excluded.
-      // The `exclude` list *must* be updated with every change to loader extensions.
-      // When adding a new loader, you must add its `test`
-      // as a new entry in the `exclude` list in the "url" loader.
+function getConfig(setName) {
+    // Note: defined here because it will be used more than once.
+    var cssFilename = setName ?
+        `static/css/${setName}.[contenthash:8].css` :
+        `static/css/[name].[contenthash:8].css`;
```
Comment (Vittly) [inline packages/bem-react-scripts/config/webpack.config.prod.js]: I think we need only two variables:

```js
const name = setName || '[name]',
    appBuild = setName ? `${appBuild}/${setName}` : appBuild;

// and use them all the way down
```

You also can write it in a hash:

```js
const target = setName ?
    { name: setName, build: `${appBuild}/${setName}` } :
    { name: '[name]', build: appBuild };

// use `target.name` and `target.build`
```

**Review by Vittly — CHANGES_REQUESTED** [review]

File: packages/bem-react-scripts/scripts/build.js (hunk 15 above)
Comment (Lakate) [reply packages/bem-react-scripts/scripts/build.js:81]: parallel-webpack gives a lot of information

File: README.md [hunk 19]
```diff
@@ -117,10 +117,19 @@ Your app is ready to be deployed!
...
+* [bem/sdk.config](https://github.com/bem/bem-sdk/blob/master/packages/config/README.md#config-example)
 * [babel-plugin-bem-import](https://github.com/bem/babel-plugin-bem-import)
 * [bem-react-core](https://github.com/bem/bem-react-core)
 * [raw-loader](https://github.com/webpack-contrib/raw-loader) includes content of *.inline.* files
 
+## Building multiple bundles
+
+If you want to make several bundles and have their own implementation of blocks, you need to use [sets](https://github.com/bem/bem-sdk/blob/master/packages/config/README.md#config-example)
+
+1. Add sets in .bemrc
```
Comment (Vittly) [inline README.md]: use backticks for `.bemrc`

File: packages/bem-react-scripts/config/webpack.config.dev.js (hunk 6 above)
Comment (Vittly) [reply packages/bem-react-scripts/config/webpack.config.dev.js:80]: he asks about `[name]` notation. May be it somehow can handle `setName`

File: packages/bem-react-scripts/config/webpack.config.prod.js [hunk 20]
```diff
@@ -42,261 +42,272 @@ if (env.stringified['process.env'].NODE_ENV !== '"production"') {
...
+                },
+                // Process JS with Babel.
+                {
+                    test: /\.(js|jsx)$/,
+                    use: [
+                        {
+                            loader: 'webpack-bem-loader',
+                            options: {
+                                levels: paths.appLevels,
+                                set: setName,
```
Comment (Vittly) [inline packages/bem-react-scripts/config/webpack.config.prod.js:180]: 👍

File: packages/bem-react-scripts/config/webpack.config.prod.js [hunk 21]
```diff
@@ -42,261 +42,272 @@ if (env.stringified['process.env'].NODE_ENV !== '"production"') {
...
+        // Some libraries import Node modules but don't use them in the browser.
+        // Tell Webpack to provide empty mocks for them so importing them works.
+        node: {
+            fs: 'empty',
+            net: 'empty',
+            tls: 'empty',
+        },
+        target: paths.appTarget,
+    };
+}
```
Comment (Vittly) [inline packages/bem-react-scripts/config/webpack.config.prod.js:309]: I hope all that stuff will work

**Comment by Vittly** [comment]
🆗

**Comment by Lakate** [comment]
@zxqfox merge? roll out a new version?

**Pull request merged**